src/arrow_format/schema.rs

```diff
@@ -0,0 +1,965 @@
+  
+  fn to_primitive_type(&self) -> Result<Option<DataType>> {
+    if self.is_self_included() {
+      self.to_primitive_type_inner().map(|dt| {
+        if self.is_repeated() {
+          Some(DataType::List(Box::new(dt)))
+        } else {
+          Some(dt)
+        }
+      })
+    } else {
+      Ok(None)
+    }
+  }
+  
+  fn to_primitive_type_inner(&self) -> Result<DataType> {
+    match self.schema.get_physical_type() {
+      PhysicalType::BOOLEAN => Ok(DataType::Boolean),
+      PhysicalType::INT32 => self.to_int32(),
+      PhysicalType::INT64 => self.to_int64(),
+      PhysicalType::FLOAT => Ok(DataType::Float32),
+      PhysicalType::DOUBLE => Ok(DataType::Float64),
+      PhysicalType::BYTE_ARRAY => self.to_byte_array(),
+      other => Err(ParquetError::ArrowError(format!("Unable to convert parquet type {}", other)))
+    }
+  }
+  
+  fn to_int32(&self) -> Result<DataType> {
+    match self.schema.get_basic_info().logical_type() {
+      LogicalType::NONE => Ok(DataType::Int32),
+      LogicalType::UINT_8 => Ok(DataType::UInt8),
+      LogicalType::UINT_16 => Ok(DataType::UInt16),
+      LogicalType::UINT_32 => Ok(DataType::UInt32),
+      LogicalType::INT_8 => Ok(DataType::Int8),
+      LogicalType::INT_16 => Ok(DataType::Int16),
+      LogicalType::INT_32 => Ok(DataType::Int32),
+      other => Err(ParquetError::ArrowError(format!("Unable to convert parquet logical type {}", other)))
+    }
+  }
+  
+  fn to_int64(&self) -> Result<DataType> {
+    match self.schema.get_basic_info().logical_type() {
+      LogicalType::NONE => Ok(DataType::Int64),
+      LogicalType::INT_64 => Ok(DataType::Int64),
+      LogicalType::UINT_64 => Ok(DataType::UInt64),
+      other => Err(ParquetError::ArrowError(format!("Unable to convert parquet logical type {}", other)))
+    }
+  }
+  
+  fn to_byte_array(&self) -> Result<DataType> {
+    match self.schema.get_basic_info().logical_type() {
+      LogicalType::UTF8 => Ok(DataType::Utf8),
+      other => Err(ParquetError::ArrowError(format!("Unable to convert parquet logical type {}", other)))
+    }
+  }
+  
+  // Functions for group types.
+  
+  fn to_struct(&self) -> Result<Option<DataType>> {
+    match &*self.schema {
+      Type::PrimitiveType {..} => panic!("This should not happen."),
+      Type::GroupType {
+        basic_info: _,
+        fields
+      } => {
+        fields.iter()
+          .map(|field_ptr| self.clone_with_schema(field_ptr.clone()).to_field())
+          .collect::<Result<Vec<Option<Field>>>>()
+          .map(|result| result.into_iter()
+            .filter_map(|f| f)
+            .collect::<Vec<Field>>())
+          .map(|fields| {
+            if fields.is_empty() {
+              None
+            } else {
+              Some(DataType::Struct(fields))
+            }
+          })
+      }
+    }
+  }
+  
+  fn to_list(&self) -> Result<Option<DataType>> {
+    match &*self.schema {
+      Type::PrimitiveType {..} => panic!("This should not happen."),
```

Comment:
- I think panic when it's primitive type is same.
- I don't think we should panic when the `fields.len() != 1` because this is an error format and should not cause our function to panic.

Reply:
You're right. For the latter maybe we can throw "not yet supported" error.

Reply:
Why can't we return Err here?

Reply:
Because if this happens, it means our code has bug.

Reply:
To start with, returning an Err does not necessarily mean the code has bug. Err is just a way of expressing an invalid state or a bug and returning it without panicking. With Err it is easier to add a unit test to show that we return a proper error message.

By the way, there is no test for this, either with panic or Err. Also, what does it mean "It should not happen". What should not happen? Can you update the error message?

Reply:
If you want to go with panic, I would suggest including something like the actual schema that was compared, and some other metadata as part of the error message, so it is easier to debug later on when it happens.